feat(deprecated module): automate the generation of deprecated modules

adomani opened the pull request:
There is a competing PR at #24917: I personally like the other one better! :smile:

I will likely close this PR soon, since the other one seems better than this one.

---

This PR takes a step further in automating deprecated modules.

The command `#create_deprecated_modules (n)? (comment)? (write)?` generates deprecated modules.

Writing
```lean
#create_deprecated_modules 5 "These files are no longer relevant"
```
looks at the `lean` files in `Mathlib` that existed `4` commits ago
(i.e. the commit that you see with `git log -5`) and that are no longer present.
It shows how the corresponding deprecated modules should look like, using
`"These files are no longer relevant"` as the (optional) comment.

If the number of commits is not explicitly used, `#create_deprecated_modules` defaults to `2`,
namely, the commit just prior to the current one.

If the message is not explicitly used, `#create_deprecated_modules` defaults to
`"Auto-generated deprecation"`.
If you wish there to be no comment, use `#create_deprecated_modules 5 ""`.

Note that the command applies the *same* comment to all the files that it generates.

Finally, if everything looks correct, adding a final `write` actually generates the files:
```lean
#create_deprecated_modules 5 "These files are no longer relevant" write
```


---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

adomani committed:
fix pretty printed deprecated module

adomani committed:
shell script

adomani committed:
lean script

adomani committed:
continuing

adomani committed:
add docs

adomani committed:
checkout deleted files

adomani committed:
just print

adomani committed:
update message

adomani committed:
Actually checkout the files

adomani committed:
progress with recreating of the deprecated files

adomani committed:
partial cleanup

adomani committed:
more in Lean and sh

adomani committed:
more syntax support

adomani committed:
docs

adomani committed:
cleanup

adomani committed:
shorten

adomani committed:
include comment and update docs

adomani committed:
strip comments

adomani committed:
comment

adomani committed:
remove bash

adomani committed:
lint

grunweg commented on scripts/create_deprecated_module.lean:
```diff
@@ -0,0 +1,175 @@
+It uses
+* `fname`, as the path of a file (which need not exist);
+* `fileContent`, as the content of `fname` (regardless of whether the file exists and what its
+  content it);
```
```suggestion
  content is);

grunweg commented on scripts/create_deprecated_module.lean:
```diff
@@ -0,0 +1,175 @@
+* `fname`, as the path of a file (which need not exist);
+* `fileContent`, as the content of `fname` (regardless of whether the file exists and what its
+  content it);
+* `keepTrailing` a boolean.
```
this comment looks unfinished (but is explained in the next sentence). Can you reword this somehow?

adomani replied:
Done!

grunweg commented on scripts/create_deprecated_module.lean:
```diff
@@ -0,0 +1,175 @@
+* `keepTrailing` a boolean.
+
+It returns the content of `fileContent` up to the final import, including trailing whitespace and
+comments if `keepTrailing = true`.
```
If I'm reading the code right, comments are always kept - but whitespace only if keepTrailing is true. Is that correct - if so, can you reword the comment, please?

adomani replied:
Regardless of what `keepTrailing` is, the *actually trailing* whitespace is removed.

However, if `keepTrailing = true` then all trailing whitespace up to and including the last comment are kept, with the understanding that "no comment" means "no whitespace" as well!  Basically, with "whitespace" I was referring to the whitespace between the strict end of the syntax and end of the final comment.  Whitespace after the last comment is always discarded.

I am not sure that explaining this in such detail is more useful than leaving as is, though.  :smile:

grunweg replied:
I like the solution you chose.

grunweg commented on scripts/create_deprecated_module.lean:
```diff
@@ -0,0 +1,175 @@
+/--
+`mkDeprecation customMessage` returns the formatted syntax
+`deprecated_module "customMessage" (since := "YYYY-MM-DD")`,
+where the message is the input and the date is today's date.
```
"where the message is the input" reads a little confusing to me - just remove that part?

adomani replied:
Removed.

grunweg commented on scripts/create_deprecated_module.lean:
```diff
@@ -0,0 +1,175 @@
+  liftCoreM <| PrettyPrinter.ppCategory `command stx
+
+/--
+The command `#create_deprecated_modules (n)? (comment)? (write)?` generates deprecated modules.
```
do you think "generates module deprecations" reads more nicely? (I don't have a strong preference.)

adomani replied:
Yes, I like it!

grunweg commented on scripts/create_deprecated_module.lean:
```diff
@@ -0,0 +1,175 @@
+-/
+--#create_deprecated_modules 10 "a comment here" --write
+
+/--
```
This reads the header of the current file to test itself. Clever :-)
Perhaps add a comment at the very top that the comment is intentional and not leftover debugging code?

adomani replied:
I like self-referentiality! :smile: 

I expanded the existing comment to mention that it is used in a test.

grunweg commented on scripts/create_deprecated_module.lean:
```diff
@@ -0,0 +1,175 @@
+| `(#create_deprecated_modules%$tk $[$nc:num]? $[$comment:str]? $[write%$write?]?) => do
+  let n := nc.getD (Syntax.mkNumLit "2") |>.getNat
+  let mut msgs := #[]
+  let getHash (n : Nat) := do
```
How about a comment, say:
`--Get the hash of the commit at `git log -n` (and throw and error if that doesn't exist).`

adomani replied:
Comment added.

grunweg commented on scripts/create_deprecated_module.lean:
```diff
@@ -0,0 +1,175 @@
+  let noFiles := onlyPastFiles.size
+  msgs := msgs.push
+    m!"{noFiles} Lean file{if noFiles == 1 then "" else "s"} in 'Mathlib' that no longer exist."
+  let deprecation← if let some cmt := comment then mkDeprecation cmt.getString else mkDeprecation
```
add a space here?

adomani replied:
Space added.

grunweg commented on scripts/create_deprecated_module.lean:
```diff
@@ -0,0 +1,175 @@
+  let pastFiles ← getFilesAtHash pastHash
+  msgs := msgs.push m!"{pastFiles.size} files at the past hash {pastHash}\n"
+  let onlyPastFiles := pastFiles.filter fun fil ↦
+    fil.takeRight ".lean".length == ".lean" &&
```
Isn't there an `endsWith` method? That seems a bit clearer to read.

adomani replied:
Thanks!

grunweg commented on scripts/create_deprecated_module.lean:
```diff
@@ -0,0 +1,175 @@
+    m!"{noFiles} Lean file{if noFiles == 1 then "" else "s"} in 'Mathlib' that no longer exist."
+  let deprecation← if let some cmt := comment then mkDeprecation cmt.getString else mkDeprecation
+  msgs := msgs.push ""
+  for fname in onlyPastFiles do
```
-- Generate a module deprecation for each file that only existed in the past commit.

adomani replied:
Comment added.

grunweg commented on scripts/create_deprecated_module.lean:
```diff
@@ -0,0 +1,175 @@
+import Mathlib.Tactic.Linter.DeprecatedModule
+-/
+#guard_msgs in
+run_cmd
```
I wonder if there's a way to test the actual deprecation output; that seems harder to arrange, though. I'm fine with just these tests!

adomani replied:
I could extract lots of small functions that perform each part of the deprecation and test them separately, but I am not sure that it would improve the script.

Right now, the script is pretty verbose in saying what it is doing and what the outcome is, so testing "live" should be straightforward.

grunweg reviewed (approved):
Wow; I would have probably gone with something much simpler/a quick shell script. And you just wrote a proper Lean command. Great to see more infrastructure being written in Lean!
I have my usual share of pernickety comments, otherwise this looks good to me. Let's merge this and try it out!

adomani committed:
Update scripts/create_deprecated_module.lean

Co-authored-by: grunweg <rothgami@math.hu-berlin.de>

adomani reviewed (commented):
Michael, thank you very much for your comments!

adomani committed:
Michael's comments

adomani committed:
Merge branch 'adomani/deprecated_module_automation' of github.com:leanprover-community/mathlib4 into adomani/deprecated_module_automation

adomani commented:
> Wow; I would have probably gone with something much simpler/a quick shell script. And you just wrote a proper Lean command. Great to see more infrastructure being written in Lean! I have my usual share of pernickety comments, otherwise this looks good to me. Let's merge this and try it out!

Writing it in Lean has the advantage that you can store more easily data in variables, which in turn helps not create intermediate files. Moreover, programming in Lean is so much easier and readable than any of the bash utilities!

adomani committed:
Merge remote-tracking branch 'origin/master' into adomani/deprecated_module_automation

adomani committed:
add `s` to filename

adomani committed:
explain better trim and comments

grunweg commented on scripts/create_deprecated_modules.lean:
```diff
@@ -0,0 +1,173 @@
+def mkDeprecation (customMessage : String := "Auto-generated deprecation") :
+    CommandElabM Format := do
+  let msgStx := if customMessage.isEmpty then none else some <| Syntax.mkStrLit customMessage
+  let dateStx := Syntax.mkStrLit s!"{← Std.Time.PlainDate.now}"
```
Is this date in the UTC time zone (or that does this depend on the computer executing it)?

adomani replied:
I think that it is, yes, but I have not done intensive testing.  I remember that the date-commands had some issues in CI, so I'm sure that there will be some joy involved in getting this to work on all setups.  This is also a reason to wait on adding the workflow to CI.

grunweg replied:
One great feature of Lean is that we have access to the source (and it's in a language I can read): [my impression](https://github.com/leanprover/lean4/blob/81b85203c90439fb7d7e0cec32130d62d8cb7a2e/src/Std/Time/Zoned.lean#L19-L28) is that this uses the time zone on the computer. Oops. Let me ask on zulip about the best way forward.

grunweg replied:
[Asked on zulip](https://leanprover.zulipchat.com/#narrow/channel/270676-lean4/topic/Current.20date.20in.20UTC/with/517985138)

grunweg replied:
```suggestion
  -- Get the current date in UTC: we don't want this to depend on the user computer's time zone.
  let dateStx := Syntax.mkStrLit s!"{(← DateTime.now (tz := .UTC)).toPlainDate}"

adomani replied:
Thanks for chasing this up!

grunweg reviewed (commented):
Thanks for the fast response. Let me add "writing in Lean means everything is documented well and there are no hidden pitfalls" (don't remember if that syntax was specific to bash? you don't need to) to the list of good reasons. Plus, actually parsing a file header is more robust than whatever shell hack I would have come up with.

One last question and then let's ship this!
maintainer delegate

adomani committed:
Update scripts/create_deprecated_modules.lean

Co-authored-by: grunweg <rothgami@math.hu-berlin.de>

adomani committed:
longline

adomani committed:
Merge branch 'adomani/deprecated_module_automation' of github.com:leanprover-community/mathlib4 into adomani/deprecated_module_automation

adomani committed:
fix

adomani committed:
better reports

Vierkantor commented:
> There is a competing PR at https://github.com/leanprover-community/mathlib4/pull/24917: I personally like the other one better! 😄

Sorry, I'm not completely sure what the description means for this PR: do you want to continue with this one simultaneously? Or should the PR description be updated?

adomani commented:
Anne, sorry for the confusion: I opened this PR first, then I tried something out in the other one and commented on this one.  Now, I think that the other PR is definitely better, but I forgot to update the comment on this one.

I now changed this one to `draft`, removed the `maintainer-merge` label and expanded the PR description with my likely intention to close this one soon.

Vierkantor commented:
Got it, thank you!
